chain/vm/fvm.go

```diff
@@ -308,6 +315,177 @@ func NewFVM(ctx context.Context, opts *VMOpts) (*FVM, error) {
 	}, nil
 }
 
+func NewDebugFVM(ctx context.Context, opts *VMOpts) (*FVM, error) {
+	state, err := state.LoadStateTree(cbor.NewCborStore(opts.Bstore), opts.StateBase)
+	if err != nil {
+		return nil, err
+	}
+
+	circToReport, err := opts.CircSupplyCalc(ctx, opts.Epoch, state)
+	if err != nil {
+		return nil, err
+	}
+
+	baseBstore := opts.Bstore
+	overlayBstore := blockstore.NewMemorySync()
+	vmBstore := blockstore.NewTieredBstore(overlayBstore, baseBstore)
+
+	fvmopts := &ffi.FVMOpts{
+		FVMVersion: 0,
+		Externs: &FvmExtern{
+			Rand:       opts.Rand,
+			Blockstore: vmBstore,
+			lbState:    opts.LookbackState,
+			base:       opts.StateBase,
+			epoch:      opts.Epoch,
+		},
+		Epoch:          opts.Epoch,
+		BaseFee:        opts.BaseFee,
+		BaseCircSupply: circToReport,
+		NetworkVersion: opts.NetworkVersion,
+		StateBase:      opts.StateBase,
+		Tracing:        EnableDetailedTracing,
+	}
+
+	fvmopts.Debug = true
+
+	// TODO move these two util functions somewhere more general
+	loadBundle := func(bundle string) (cid.Cid, error) {
+		f, err := os.Open(bundle)
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("error opening debug bundle: %w", err)
+		}
+		defer f.Close() //nolint
+
+		hdr, err := car.LoadCar(ctx, overlayBstore, f)
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("error loading debug bundle: %w", err)
+		}
+
+		return hdr.Roots[0], nil
+	}
+
+	loadManifest := func(mfCid cid.Cid) (manifest.ManifestData, error) {
+		adtStore := adt.WrapStore(ctx, cbor.NewCborStore(overlayBstore))
+
+		var mf manifest.Manifest
+		var mfData manifest.ManifestData
+
+		if err := adtStore.Get(ctx, mfCid, &mf); err != nil {
+			return mfData, xerrors.Errorf("error reading debug manifest: %w", err)
+		}
+
+		if err := mf.Load(ctx, adtStore); err != nil {
+			return mfData, xerrors.Errorf("error loading debug manifest: %w", err)
+		}
+
+		if err := adtStore.Get(ctx, mf.Data, &mfData); err != nil {
+			return mfData, xerrors.Errorf("error fetching manifest data: %w", err)
+		}
+
+		return mfData, nil
+	}
+
+	putMapping := func(ar map[cid.Cid]cid.Cid) (cid.Cid, error) {
+		var mapping xMapping
+
+		mapping.redirects = make([]xRedirect, 0, len(ar))
+		for from, to := range ar {
+			mapping.redirects = append(mapping.redirects, xRedirect{from: from, to: to})
+		}
+		sort.Slice(mapping.redirects, func(i, j int) bool {
+			return bytes.Compare(mapping.redirects[i].from.Bytes(), mapping.redirects[j].from.Bytes()) < 0
+		})
+
+		// Passing this as a pointer of structs has proven to be an enormous PiTA; hence this code.
+		cborStore := cbor.NewCborStore(overlayBstore)
+		mappingCid, err := cborStore.Put(context.TODO(), &mapping)
+		if err != nil {
+			return cid.Undef, err
+		}
+
+		return mappingCid, nil
+	}
+
+	switch opts.NetworkVersion {
```

Comment:
This is technically wrong -- we should decide based on actors version, not network version. You can use [this method](https://github.com/filecoin-project/lotus/blob/3370847f3e98ed1c4c6c247a01db863b5bddd4a9/chain/actors/version.go#L41).

Reply:
nitpicker! ok, will fix.

Reply:
done.